src/generic/gridsel.cpp

```diff
@@ -975,222 +572,330 @@ void wxGridSelection::ClearSelection()
+        return block.GetBottomRow();
+    if ( block.GetBottomRow() == m_grid->m_currentCellCoords.GetRow() )
+        return block.GetTopRow();
+
+    return -1;
+}
+
+int wxGridSelection::GetCurrentBlockCornerCol() const
+{
+    if ( m_selection.empty() )
+        return -1;
+
+    const wxGridBlockCoords& block = *m_selection.rbegin();
+    if ( block.GetLeftCol() == m_grid->m_currentCellCoords.GetCol() )
+        return block.GetRightCol();
+    if ( block.GetRightCol() == m_grid->m_currentCellCoords.GetCol() )
+        return block.GetLeftCol();
+
+    return -1;
+}
+
+wxGridCellCoordsArray wxGridSelection::GetCellSelection() const
+{
+    if ( m_selectionMode != wxGrid::wxGridSelectCells )
+        return wxGridCellCoordsArray();
+
+    wxGridCellCoordsArray cells;
+    const size_t count = m_selection.size();
+    for ( size_t n = 0; n < count; n++ )
+    {
+        const wxGridBlockCoords& block = m_selection[n];
+        if ( block.GetTopRow() == block.GetBottomRow() &&
+             block.GetLeftCol() == block.GetRightCol() )
+        {
+            cells.Add(block.GetTopLeft());
+        }
+    }
+    return cells;
+}
+
+wxGridCellCoordsArray wxGridSelection::GetBlockSelectionTopLeft() const
+{
+    // return blocks only in wxGridSelectCells selection mode
+    if ( m_selectionMode != wxGrid::wxGridSelectCells )
+        return wxGridCellCoordsArray();
+
+    wxGridCellCoordsArray coords;
+    const size_t count = m_selection.size();
+    for ( size_t n = 0; n < count; n++ )
+    {
+        coords.Add(m_selection[n].GetTopLeft());
+    }
+    return coords;
+}
+
+wxGridCellCoordsArray wxGridSelection::GetBlockSelectionBottomRight() const
+{
+    if ( m_selectionMode != wxGrid::wxGridSelectCells )
+        return wxGridCellCoordsArray();
+
+    wxGridCellCoordsArray coords;
+    const size_t count = m_selection.size();
+    for ( size_t n = 0; n < count; n++ )
+    {
+        coords.Add(m_selection[n].GetBottomRight());
+    }
+    return coords;
+}
+
+wxArrayInt wxGridSelection::GetRowSelection() const
+{
+    if ( m_selectionMode == wxGrid::wxGridSelectColumns )
+        return wxArrayInt();
+
+    wxIntSet uniqueRows;
+    const size_t count = m_selection.size();
+    for ( size_t n = 0; n < count; ++n )
+    {
+        const wxGridBlockCoords& block = m_selection[n];
+        if ( block.GetLeftCol() == 0 &&
+             block.GetRightCol() == m_grid->GetNumberCols() - 1 )
+        {
+            for ( int r = block.GetTopRow(); r <= block.GetBottomRow(); ++r )
+            {
+                uniqueRows.insert(r);
+            }
+        }
+    }
+
+    wxArrayInt result;
+    result.Alloc(uniqueRows.size());
+    wxIntSet::iterator it;
+    for( it = uniqueRows.begin(); it != uniqueRows.end(); ++it )
+    {
+        result.Add(*it);
+    }
+    return result;
+}
+
+wxArrayInt wxGridSelection::GetColSelection() const
+{
+    if ( m_selectionMode == wxGrid::wxGridSelectRows )
+        return wxArrayInt();
+
```

Comment:
I think some comment describing what's going on here would be welcome, e.g.
```
// For compatibility with the existing code, try reconstructing the selected
// columns from the selected blocks we store internally. Of course, this only
// works well in the column selection mode in which the user can only select
// the entire columns in the first place, as otherwise it's difficult to
// efficiently determine that a column is selected because all of its cells
// were selected one by one. But this should work well enough in practice and
// is, anyhow, the best we can do.
```
Perhaps you could put this comment before both row/column selection functions to avoid repeating it.